File: packs/stack/python-fastapi/scaffold/app/domain/workflows/notes_workflow.py
import uuid

from app.core.exceptions import NotFoundError
from app.domain.entities.note import Note
from app.domain.interfaces.note_repository import NoteRepository


class NotesWorkflow:
    def __init__(self, repository: NoteRepository) -> None:
        self._repository = repository

    async def create_note(self, title: str, body: str = "") -> Note:
        note = Note.create(note_id=uuid.uuid4().hex, title=title, body=body)
        return await self._repository.add(note)

    async def get_note(self, note_id: str) -> Note:
        note = await self._repository.get(note_id)
        if note is None:
            raise NotFoundError("note", note_id)
        return note

    async def list_notes(self) -> list[Note]:
        return await self._repository.list_all()

File: packs/stack/typescript-node/scaffold/src/workflows/notes-workflow.ts
import { randomUUID } from "node:crypto";
import { z } from "zod";
import { DomainValidationError, NotFoundError } from "../core/errors.js";
import { createNote, type Note } from "../domain/note.js";
import type { NoteRepository } from "../domain/note-repository.js";

/**
 * Orchestrates the notes use cases. Routes call these methods and nothing else reaches around
 * them to touch a repository. Library failures are translated into domain errors here, so the
 * layers above never have to know that zod produced them.
 */
export class NotesWorkflow {
  readonly #repository: NoteRepository;

  constructor(repository: NoteRepository) {
    this.#repository = repository;
  }

  async createNote(title: string, body = ""): Promise<Note> {
    let note: Note;
    try {
      note = createNote({ id: randomUUID().replaceAll("-", ""), title, body });
    } catch (error) {
      if (error instanceof z.ZodError) {
        throw new DomainValidationError(error.issues[0]?.message ?? "invalid note");
      }
      throw error;
    }
    return await this.#repository.add(note);
  }

  async getNote(id: string): Promise<Note> {
    const note = await this.#repository.get(id);
    if (note === null) throw new NotFoundError("note", id);
    return note;
  }

  async listNotes(): Promise<Note[]> {
    return await this.#repository.listAll();
  }
}
